File: CopasiCS/CPP/Client/Message.cpp
#include <iostream>
#include <string>
#include <string.h>
#include <vector>
#include <stdio.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "Message.h"

#define BUFFER_SIZE 1000
#define MESSAGE_HEADER_SIZE 3


//Initializing static members
Mutex Message::readMutex = Mutex();
Mutex Message::writeMutex = Mutex();

Message::Message()
{
    uniqueID = Constants::MESSAGE_START;
    runningObject = Constants::MESSAGE_STOP;
    messageType = Constants::MESSAGE_PAUSE;
    noOfElements = 0;
    //parameters;
    fullSimulationSpec = "FSS";

    partialSimulationSpec = "PSS";
    errorMessage = "EM";
}


void Message::readObject(int in)
{

    Message::readMutex.lock();

    showMessage("Inside readObject: " );
    cerr << in << endl;

    /*cerr << endl << "Header: " << endl;
      for (int i = 0; i < 0; i++)
      {
      read(in, &ch, sizeof(ch));
      cerr << "Char " << i << ": " << ch <<" " << (int)ch << '\t';
      }*/


    uniqueID = this->readInt(in);
    //showMessage("Read uniqueID" );
    cerr << uniqueID << endl;




    runningObject = this->readInt(in);
    //showMessage("Read RunningObject" );
    cerr << runningObject << endl;

    messageType = this->readInt(in);
    //showMessage("Read MessageType" );
    cerr << messageType << endl;

    noOfElements = this->readInt(in);
    //showMessage("Read noOfElements" );
    cerr << noOfElements << endl;

    for (int i = 0; i < noOfElements; i++)
    {
#ifdef __MACHINE_IS_SMALL_ENDIAN
        double tmpDouble = this->readDoubleForSmallEndianMachines(in);
#else
        double tmpDouble = this->readDouble(in);
#endif


        //this->parameters.push_back(ntohd(tmpDouble));
        cerr << "Read double: " << tmpDouble << endl;
        this->parameters.push_back(tmpDouble);
    }


    char *tmpCharBuffer;
    int strLen;

    strLen = this->readInt(in);
    tmpCharBuffer = new char[strLen + 1];
    read(in, tmpCharBuffer, strLen);
    *(tmpCharBuffer + strLen) = '\0';
    fullSimulationSpec = tmpCharBuffer;
    delete [] tmpCharBuffer;




    /*
      tmpInt = 0;
      cerr<< "Reading " << sizeof(tmpInt) << " chars..." << endl;
      cerr << "Chars are: " << endl;
      //read(in, &tmpInt, sizeof(tmpInt));
      for (unsigned int i = 0; i < sizeof(tmpInt); i++)
      {
      read(in, &ch, sizeof(ch));
      cerr << "Char " << i << ": " << ch <<" " << (int)ch << '\t';
      *(p + i) = ch;
      }

      cerr << endl << "String Length: " << tmpInt << endl;
      strLen = ntohl(tmpInt);
      cerr << endl << "String Length: " << strLen << endl;

      tmpCharBuffer = new char[strLen + 1];
      read(in, tmpCharBuffer, strLen);
      *(tmpCharBuffer + strLen) = '\0';

      for (int i = 0; i < strLen; i++)
      {
      cerr << "Char " << i << ": " << *(tmpCharBuffer + i) << " " << (int) *(tmpCharBuffer + i)<< '\t';
      if (i > 100)
      break;
      }
      cerr << endl << "String Read: " << tmpCharBuffer << endl;
    */

    strLen = this->readInt(in);
    tmpCharBuffer = new char[strLen + 1];
    read(in, tmpCharBuffer, strLen);
    *(tmpCharBuffer + strLen) = '\0';
    partialSimulationSpec = tmpCharBuffer;
    delete [] tmpCharBuffer;


    strLen = this->readInt(in);
    tmpCharBuffer = new char[strLen + 1];
    read(in, tmpCharBuffer, strLen);
    *(tmpCharBuffer + strLen) = '\0';
    errorMessage = tmpCharBuffer;
    delete [] tmpCharBuffer;

    Message::readMutex.unlock();

}


void Message::writeObject(int out)
{
    Message::writeMutex.lock();

    showMessage("Inside writeObject: ");
    cerr << out << endl;
    int tmpInt;
    tmpInt = htonl(uniqueID);
    write(out, &tmpInt, sizeof(tmpInt));
    tmpInt = htonl(runningObject);
    write(out, &tmpInt, sizeof(tmpInt));
    tmpInt = htonl(messageType);
    write(out, &tmpInt, sizeof(tmpInt));

    tmpInt = htonl(noOfElements);
    write(out, &tmpInt, sizeof(tmpInt));

    for (int i = 0; i < noOfElements; i++)
    {
        //double tmpDouble = htond(this->parameters[i]);
        double tmpDouble = this->parameters[i];
        cerr << "Writing Double: " << tmpDouble << endl;

#ifdef __MACHINE_IS_SMALL_ENDIAN
        this->writeDoubleForSmallEndianMachines(out, tmpDouble);
#else
        write(out, &tmpDouble, sizeof(tmpDouble));
#endif

    }


    char *tmpCharBuffer;
    int strLen;

    strLen = fullSimulationSpec.length();
    tmpInt = htonl(strLen);
    write(out, &tmpInt, sizeof(tmpInt));
    tmpCharBuffer = new char[strLen];
    strcpy(tmpCharBuffer, fullSimulationSpec.c_str());
    write(out, tmpCharBuffer, strLen);
    delete [] tmpCharBuffer;


    strLen = partialSimulationSpec.length();
    tmpInt = htonl(strLen);
    write(out, &tmpInt, sizeof(tmpInt));
    tmpCharBuffer = new char[strLen];
    strcpy(tmpCharBuffer, partialSimulationSpec.c_str());
    write(out, tmpCharBuffer, strLen);
    delete [] tmpCharBuffer;


    strLen = errorMessage.length();
    tmpInt = htonl(strLen);
    write(out, &tmpInt, sizeof(tmpInt));
    tmpCharBuffer = new char[strLen];
    strcpy(tmpCharBuffer, errorMessage.c_str());
    write(out, tmpCharBuffer, strLen);
    delete [] tmpCharBuffer;

    Message::writeMutex.unlock();
}



string Message::printMessageDetails()
{
    string temp;
    temp.append("\nThe message fields are as follows: ");
    temp.append("\nUnique ID: ");
    char buf[BUFFER_SIZE];
    sprintf(buf, "%d", this->uniqueID);
    temp.append(buf);
    temp.append("\nRunning Object: ");
    sprintf(buf, "%d", this->runningObject);
    temp.append(buf);
    temp.append("\nMessage Type: ");
    sprintf(buf, "%d", this->messageType);
    temp.append(buf);
    temp.append("\nNumber Of Elements in vector: ");
    sprintf(buf, "%d", this->noOfElements);
    temp.append(buf);
    temp.append("\nVector Elements are: ");
    for (int i = 0; i < this->noOfElements; i++)
    {
        double tmpDouble = (double) parameters[i];
        temp.append("\n\tParameter ");
        sprintf(buf, "%d", i);
        temp.append(buf);
        temp.append(": ");
        sprintf(buf, "%f", tmpDouble);
        temp.append(buf);
    }
    temp.append("\nFull Simulation Specification: ");
    //temp += fullSimulationSpec;
    temp.append(this->fullSimulationSpec.c_str());

    temp.append("\nPartial Simulation Specification: ");
    //temp += partialSimulationSpec;
    temp.append(this->partialSimulationSpec.c_str());

    temp.append("\nError Message: ");
    //temp += errorMessage;
    temp.append(this->errorMessage.c_str());

    return temp;
}


void Message::readMessageFromKeyboard()
{
    showMessage("\nEnter the message fields: ");
    cout << "Unique ID: ";

    cin >> this->uniqueID;
    cout << "Running Object: ";

    cin >> this->runningObject;


    cout << "Message Type: ";
    cin >> this->messageType;


    cout << "Enter Vector Elements (Doubles): ";
    string choice = "Y";
    int i = 1;
    while ((choice.substr(0, 1) == "y") || choice.substr(0, 1) == "Y")
    {
        double tmpDouble = 0.0;
        cout << "Enter Vector Element " << i << ": ";

        cin >> tmpDouble;
        this->parameters.push_back(tmpDouble);
        //this->parameters.insert(i-1, tmpDouble);
        cout << "\tAdd Another (Y/N): ";
        cin >> choice;
        i++;
        this->noOfElements++;
    }
    //this->noOfElements = this->parameters.size();

    cout << "Full Simulation Specification : ";
    cin >> this->fullSimulationSpec;

    cout << "Partial Simulation Specification : ";
    cin >> this->partialSimulationSpec;

    cout << "Error Message: ";
    cin >> this->errorMessage;
}


/** Reads a (4 byte) integer from a stream given as in.
*   Result returned as int after Network to Host Conversion.
*/

int Message::readInt(int in)
{
    int tmpInt = 0;
    char *p = (char *) &tmpInt;
    char ch;
    for (unsigned int i = 0; i < sizeof(tmpInt); i++)
    {
        read(in, &ch, sizeof(ch));
        //cerr << "Char " << i << ": " << ch <<" " << (int)ch << '\t';
        *(p + i) = ch;
    }
    return (ntohl(tmpInt));
}

#ifdef __MACHINE_IS_SMALL_ENDIAN
double Message::readDoubleForSmallEndianMachines(int in)
{
    double tmpDouble = 0.0;
    char *p = (char *) &tmpDouble;
    char ch;
    for (int i = (sizeof(tmpDouble) - 1); i >= 0; i--)
    {

        read(in, &ch, sizeof(ch));
        cerr << "Char " << i << ": " << ch <<" " << (int)ch << '\t';
        *(p + i) = ch;
    }
    //return (ntohl(tmpInt));
    //Double conversion(ntohd) to do
    return (tmpDouble);
}

void Message::writeDoubleForSmallEndianMachines(int out, double doubleToWrite)
{
    char *p = (char *) &doubleToWrite;
    char ch;
    for (int i = (sizeof(doubleToWrite) - 1); i >= 0 ; i--)
    {
        ch = *(p + i);
        write(out, &ch, sizeof(ch));
    }
}

#else
double Message::readDouble(int in)
{
    double tmpDouble = 0.0;
    char *p = (char *) &tmpDouble;
    char ch;
    for (unsigned int i = 0; i < sizeof(tmpDouble); i++)
    {

        read(in, &ch, sizeof(ch));
        cerr << "Char " << i << ": " << ch <<" " << (int)ch << '\t';
        *(p + i) = ch;
    }
    //return (ntohl(tmpInt));
    //Double conversion(ntohd) to do
    return (tmpDouble);
}
#endif

File: CopasiCS/Java/Common/Message.java
package Common;

import java.lang.*;
import java.util.*;
import java.io.*;

/**The data members of this class are the fields which 
 * are to be passed between the client and the server.
 */
public class Message implements Serializable
{
	/**The ID of this message. */
	public int uniqueID;
	/**Running object ID. */
	public int runningObject;
	/**Type of message*/
	public int messageType;
	/**Number of Elements that the vector (of doubles) contains.*/
	public int noOfElements;
	/**The vector containing the doubles which would be 
	 * passed as parameters to the corresponding function.
	 */
	public Vector parameters;
	/**The full simulation specification.*/
	public String fullSimulationSpec;
	/**The partial simulation specification.*/
	public String partialSimulationSpec;
	/**Error message if there is an error*/
	public String errorMessage;
	

	/**No argument constructor. Sets up the default values for the
	 * different fields.
	 */
	public Message()
	{
		
        uniqueID = Constants.MESSAGE_START;
		runningObject = 0;
		messageType = 0;
		noOfElements = 0;
		parameters = new Vector();
		fullSimulationSpec = null;
		String partialSimulationSpec = null;
		String errorMessage = null;
				
	}
	

 	/**Writes the fields of this message to the socket through
 	 * a DataOutputStream. Locks the DataOutputStream before
 	 * writing so that some other thread does not corrupt the
 	 * stream.
 	 * @param out The DataOutputStream constructed from the 
 	 * socket output stream. 
 	 */
	public void writeDataObject(java.io.DataOutputStream out) throws IOException
 	{
 			
 		synchronized (out)
 		{
	 		out.writeInt(uniqueID);
			out.writeInt(runningObject);
			out.writeInt(messageType);
			
			out.writeInt(noOfElements);
			for (int i = 0; i < noOfElements; i++)
			{
				Double tmpDouble = (Double) this.parameters.get(i);
				out.writeDouble(tmpDouble.doubleValue());
			}
			
			out.writeInt(fullSimulationSpec.length());
			out.writeBytes(fullSimulationSpec);
			
			out.writeInt(partialSimulationSpec.length());
			out.writeBytes(partialSimulationSpec);
	 		
	 		
	 		out.writeInt(errorMessage.length());
	 		//showMessage("Writing Length: "  + errorMessage.length());
	 		out.writeBytes(errorMessage);
	 	}
 		 		
  	}
 	
 	
 	/**Reads the fields for this message from the socket through
 	 * a DataInputStream. Locks the DataInputStream before
 	 * writing so that some other thread does not corrupt the
 	 * stream.
 	 * @param in The DataInputStream constructed from the 
 	 * socket input stream. 
 	 */
	public void readDataObject(java.io.DataInputStream in) throws IOException, ClassNotFoundException
 	{
 		
 		synchronized (in)
 		{
	 		this.uniqueID = in.readInt();
			this.runningObject = in.readInt();
			this.messageType = in.readInt();
			
			this.noOfElements = in.readInt();
			this.parameters = new Vector(noOfElements);
			for (int i = 0; i < noOfElements; i++)
			{
				Double tmpDouble = new Double(in.readDouble());
				this.parameters.add(tmpDouble);
			}
				
			int tmpSize;
			byte[] byteString = null;
					
			tmpSize = in.readInt();
			byteString = new byte[tmpSize];
	 		for (int i = 0; i < tmpSize; i++)
	 		{
	 			byteString[i] = in.readByte();
	 		}
	 		this.fullSimulationSpec = new String(byteString);
	 		
	 		
	 		tmpSize = in.readInt();
			byteString = new byte[tmpSize];
	 		for (int i = 0; i < tmpSize; i++)
	 		{
	 			byteString[i] = in.readByte();
	 		}
	 		this.partialSimulationSpec = new String(byteString);
	 				
	 		
	 		
	 		tmpSize = in.readInt();
			byteString = new byte[tmpSize];
	 		for (int i = 0; i < tmpSize; i++)
	 		{
	 			byteString[i] = in.readByte();
	 		}
	 		this.errorMessage = new String(byteString);
	 	}
		
 	}
 		
 
	/**Makes a string containing the values of all the fields 
	 * of this message. 
	 * @return The string containing the values of all the fields
	 * of this message.
	 */
	public String printMessageDetails()
	{
		StringBuffer temp = new StringBuffer();
		temp.append("\nThe message fields are as follows: ");
		temp.append("\nUnique ID: " + this.uniqueID);
		temp.append("\nRunning Object: " + this.runningObject);
		temp.append("\nMessage Type: " + this.messageType);
		temp.append("\nNumber Of Elements in Vector: " + this.noOfElements);
		temp.append("\nVector Elements are: ");
		for (int i = 0; i < this.parameters.size(); i++)
		{
			Double tmpDouble = (Double) this.parameters.get(i);
			temp.append("\n\tParameter " + i + ": " + tmpDouble.toString());
		}
		
		temp.append("\nFull Simulation Specification: " + this.fullSimulationSpec);
		temp.append("\nPartial Simulation Specification: " + this.partialSimulationSpec);
		temp.append("\nError Message: " + this.errorMessage);
	
		return temp.toString();
	}
	
	
	/**Reads the values from the keyboard 
	 * and assigns them to the corresponding 
	 * fields of this message.
	 */
	public void readMessageFromKeyboard()
	{
		String line = null;
      	
		showMessage("\nEnter the message fields: ");
		System.out.println("Unique ID: ");
		line = 	getLine();
		try
		{
			this.uniqueID = Integer.parseInt(line);
		}
		catch(NumberFormatException nfe)
		{
			showMessage("Unique ID not proper Integer");
			uniqueID = 0;
		}
		
		System.out.println("Running Object: ");
		line = 	getLine();
		try
		{
			this.runningObject = Integer.parseInt(line);
		}
		catch(NumberFormatException nfe)
		{
			showMessage("Running Object not proper Integer");
			this.runningObject = 0;
		}
		
		System.out.println("Message Type: ");
		line = 	getLine();
		try
		{
			this.messageType = Integer.parseInt(line);
		}
		catch(NumberFormatException nfe)
		{
			showMessage("Message Type not proper Integer");
			this.messageType = 0;
		}
		
				
		System.out.println("Enter Vector Elements (Doubles): ");
		String choice = "Y";
		int i = 1;
		while (choice.trim().equals("y") || choice.trim().equals("Y"))
		{
			double tmpDouble = 0.0;
			System.out.println("Enter Vector Element " + i + ": ");
			line = 	getLine();
			try
			{
				tmpDouble = Double.parseDouble(line);
			}
			catch(NumberFormatException nfe)
			{
				showMessage("Running Object not proper Double");
			}
			
			this.parameters.add(new Double(tmpDouble));
			System.out.println("\tAdd Another (Y/N): " );
			choice = getLine();
			i++;
		}
		this.noOfElements = this.parameters.size();
		
		System.out.println("Full Simulation Specification : ");
		this.fullSimulationSpec = getLine();
		
		System.out.println("Partial Simulation Specification : ");
		this.partialSimulationSpec = getLine();
		
		System.out.println("Error Message: ");
		this.errorMessage  = getLine();
	}

		
	
	/**A small function to display messages.
	 * @param msg The message string to show.
	 */
	private void showMessage(String msg)
	{
		System.out.println("Message: " + msg);
	}
	
	/**Gets a line from the keyboard.
	 * @return The line input by the user from the keyboard
	 */
	private String getLine()
	{
		BufferedReader keyboard = new BufferedReader(new InputStreamReader(System.in));
		String line = null;
		try
		{
			line = keyboard.readLine();
		}
		catch(IOException ioe)
		{
			showMessage("Error reading from keyboard.");
		}
		return line;
	}
}
